File: src/BaseHmiTypes/Screens/Widgets/HmiSelectionGroupBase.cs
using BaseHmiTypes.Screens.Base;

namespace BaseHmiTypes.Screens.Widgets;

public abstract class HmiSelectionGroupBase : HmiWidgetBase
{
    public IList<HmiState> States { get; } = new List<HmiState>();

    /// <summary>
    /// Gets or sets the number of configured states, excluding any error state.
    /// </summary>
    public HmiProperty<int>? ConfiguredStateCount { get; set; }

    public HmiProperty<HmiStateTriggerMode>? StateTriggerMode { get; set; }

    public HmiProperty<double>? Value { get; set; }

    public IList<HmiSelectionGroupItem> Items { get; } = new List<HmiSelectionGroupItem>();

    public HmiProperty<int>? SelectedIndex { get; set; }

    public HmiProperty<int>? SelectionMode { get; set; }

    public HmiProperty<bool>? WrapAround { get; set; }

    public HmiProperty<bool>? WriteOnEnter { get; set; }

    public HmiEnterHandshakeSettings? EnterHandshake { get; set; }

    public HmiProperty<double>? SelectionItemHeight { get; set; }

    public HmiProperty<HmiColor>? SelectionBackgroundColor { get; set; }

    public HmiProperty<HmiColor>? SelectionForegroundColor { get; set; }

    public HmiProperty<HmiColor>? EvenRowBackgroundColor { get; set; }

    public HmiProperty<HmiColor>? SelectionBorderColor { get; set; }

    public HmiProperty<double>? SelectionBorderWidth { get; set; }
}

public class HmiSelectionGroupItem
{
    public string? Text { get; set; }

    public HmiImageSource? Image { get; set; }

    public string? ImageName { get; set; }
}

File: typescript/base-hmi-types/src/screens/widgets/HmiSelectionGroupBase.ts
import { HmiColor, HmiImageSource, HmiProperty } from "../base.js";
import { HmiWidgetBase } from "./HmiWidgetBase.js";
import { HmiState } from "./HmiState.js";
import { HmiEnterHandshakeSettings } from "./HmiEnterHandshakeSettings.js";
import { HmiStateTriggerMode } from "./HmiStateTriggerMode.js";

export interface HmiSelectionGroupItem {
  text?: string;
  image?: HmiImageSource;
  imageName?: string;
}

export abstract class HmiSelectionGroupBase extends HmiWidgetBase {
  readonly states: HmiState[] = [];
  configuredStateCount?: HmiProperty<number>;
  stateTriggerMode?: HmiProperty<HmiStateTriggerMode>;
  value?: HmiProperty<number>;
  /** Read-only runtime value that selects a state while remotely operated. */
  indicator?: HmiProperty<number>;
  readonly items: HmiSelectionGroupItem[] = [];
  selectedIndex?: HmiProperty<number>;
  selectionMode?: HmiProperty<number>;
  wrapAround?: HmiProperty<boolean>;
  writeOnEnter?: HmiProperty<boolean>;
  /** A nonzero read-only runtime value enables remote control; zero leaves operator control. */
  remoteAccess?: HmiProperty<number>;
  /** Write target receiving the first visible state's value when the list scrolls. */
  topPosition?: HmiProperty<number>;
  /** Array write target receiving all visible state values when the list scrolls. */
  visibleStates?: HmiProperty<readonly number[]>;
  enterHandshake?: HmiEnterHandshakeSettings;
  selectionItemHeight?: HmiProperty<number>;
  selectionBackgroundColor?: HmiProperty<HmiColor>;
  selectionForegroundColor?: HmiProperty<HmiColor>;
  evenRowBackgroundColor?: HmiProperty<HmiColor>;
  selectionBorderColor?: HmiProperty<HmiColor>;
  selectionBorderWidth?: HmiProperty<number>;
}
